__Author: Nitesh Turaga, May 6th 2014__


Goal: from Affymetrix raw data stored in CEL files to differential gene expression
========================================================



### Some Basics

Getting  the working directory, in my case

```{r getwd,highlight=TRUE,comment=NA}
getwd()
````

Setting the working directory, in my case:
```{r setwd,eval=TRUE}
setwd("~/Documents/GeneExpressionDataAnalysis/affyData/")
````

Check the date

```{r checkDate,highlight=TRUE,comment=NA}
date()
```



Bioconductor
--------------


We are going to use the following packages from bioconductor:

1. affy 
2. limma 
3. hgu95a.db
4. annotate



### To install the missing packages from Bioconductor use biocLite()

First you need to get the list of available packages
```{r installBioconductorPackages,echo=FALSE,comment=NA}
installedPckgs <- installed.packages()[,"Package"]

# Here we define the list of desired libraries
pckgListBIOC <- c("affy", "limma", "hgu95a.db", "annotate")

# We source the biocLite.R script from the Bioconductor website
source("http://bioconductor.org/biocLite.R")

# Load the packages, or install them from Bioconductor if needed
for (pckg in pckgListBIOC) {
    if (! pckg %in% installedPckgs) biocLite(pckg)
    require(pckg, character.only=TRUE)
}
```

Clear the workspace, note the two "embedded" functions

```{r clearWorkSpace,comment=NA}
rm(list=ls())
```

### Load the libraries needed for the analysis, note the use of "require"
```{r loadLibraries}
require(affy)
require(limma)
require(hgu95a.db)
require(annotate)
```


Reading and preprocessing the expression data
------------------------------------------------

Load data from CEL files using the ReadAffy function
```{r readAffyData,cache=TRUE}
dat <- ReadAffy(celfile.path="./data/raw", compress=TRUE)
```

Let's have a look at the dat object, so just print. Remember that in R, you can just enter the name of the object
and it automatically prints.
```{r printData}
# print(dat) is the same as "dat".
dat
```

Check the **Class** of the dat object
```{r dataClass,comment=NA}
class(dat)
```

You can find information about this class as usual
```{r affyBatchHelp, echo=FALSE,comment=NA}
?AffyBatch
```

Check the Dimensions of the object, the structure and information about the annotation of the data. It is important to know how the data is organized, and get some information from the **AffyBatch** class.
```{r checkDataProperties,eval=TRUE,comment=NA}
# Dimensions
dim(dat)
```

```{r comment=NA}
#Structure
str(dat)
```

```{r comment=NA}
# Information about the platform
annotation(dat)
```

```{r comment=NA,eval=FALSE}
# Information about the features
head(featureData(dat))
```

```{r comment=NA}
# Information about the samples, also called the phenotype data.
pData(dat)
```


Read the phenotype data associated with the experiment. __Note the use of colClasses__ in the following code.
colClasses, is setting each column to be read in with a specific data type. 
```{r readTargetsFile}
pheno <- read.table('./data/targets.txt',sep='\t',header=T,
      	     colClasses=c("character", "character",
	     			           "factor","factor","character"))

```

Let's have a look into the phenotypic information, and summarize some useful information.

```{r checkPhenotypeDataProperties,fig.height=5,fig.width=8,comment=NA}
str(pheno)
# Let's get summarize the data
summary(pheno)
# Factor are very useful!
table(pheno$CancerType, pheno$CellType)

plot(pheno$CancerType, pheno$CellType)
```

We can add phenotypic information to the AffyBatch object
```{r mergePhenoData,comment=NA}
if(all(rownames(pData(dat))==pheno$Name)){
	print("You're good, go ahead!")
	pheno <- merge(pData(dat), pheno, by.x=0, by.y='Name', sort=FALSE)
	rownames(pheno) <- pheno[,"Row.names"]
	pData(dat) <- pheno
}else{
	print("Check order of rows in pData(object) and pheno data provided!")
}

# Now pData(dat) is the merged object
pData(dat)
```

Normalize
----------

We can now normalize the data, for instance the  **rma()** function, Alternatives you want to look into are **frma()** and **gcrma()**.
```{r normalizeData,comment=NA}
dat.rma <- rma(dat)

# Check the new object after normalization
class(dat.rma)
```

Nevertheless the following methods still work, since **ExpressionSet is also derived from eSet, just like AffyBatch**. 
```{r checkNormalizedData,eval=FALSE}
pData(dat.rma)
annotation(dat.rma)
head(featureData(dat.rma))
dim(dat.rma)
```

Now we can save the two objects before going ahead

```{r saveData,eval =TRUE}
save(dat, dat.rma, file="./objs/affyData.rda")
```


Differential Gene Expression Analysis
==========================================

Design Matrix
-------------

To create the design matrix and to fit the linear model we start combining all levels of our factors of interest.
```{r designMatrix,comment=NA}
groups <- factor(paste(pheno$CancerType, pheno$CellType, sep="."))

# Count the number of sample per group
table(groups)

# Create the design matrix
dMat <- model.matrix(~0+groups)
colnames(dMat) <- levels(groups)

# This is the design matrix
dMat

# In reality we could just use the informations in the pheno$CancerType factor, which is equivalent
table(pheno$CancerType)
dMat2 <- model.matrix(~0+pheno$CancerType)
colnames(dMat2) <- levels(pheno$CancerType)

# This is the second design matrix
dMat2

# As you can see they are exactly the same
dMat==dMat2
```

Contrasts Matrix
------------------

Let's create now a contrast matrix to extract differences between coefficients.

A contrast matrix using the first design matrix
```{r contrastsMatrix,comment=NA}
cMat <- makeContrasts(levels=colnames(dMat),
		 CellType=( ((colon.Epithelial + breast.Epithelial)/2) -
		 	    		       	    ((Burkitt.Lymphocytes + DLBCL.Lymphocytes)/2) ),
						    			     CancerType.Epithelial=( colon.Epithelial - breast.Epithelial ),
									     			     		       CancerType.Lyphocytes=( Burkitt.Lymphocytes-DLBCL.Lymphocytes )
														       			       					      )

# All coefficients that were combined contribute equally. The columns sum must be equal to 0.
cMat
apply(cMat, 2, sum)


# A contrast matrix using the second design matrix
cMat2 <- makeContrasts(levels=colnames(dMat2),
  	       CellType=( ((colon + breast)/2) - ((Burkitt + DLBCL)/2) ),
		       CancerType.Epithelial=( colon - breast ),
		       CancerType.Lyphocytes=( Burkitt - DLBCL )
			 )

# Also in this case all the coefficients that were combined must contribute equally. The columns sum again must be  equal to 0.
cMat2
apply(cMat2, 2, sum)

# The two contrasts matrices are the same
cMat==cMat2

```


Fit the model using the first design and contrast matrices
-----------------------------------------------------------

You might want to learn about __lmFit()__ and __contrasts.fit()__ 
```{r linearModelFitandEBayes,comment=NA}
fit.ls <- lmFit(dat.rma, dMat, method="ls")
fit.ls <- contrasts.fit(fit.ls, cMat)

# Fit the model using the second design matrix and contrast matrices
fit2.ls <- lmFit(dat.rma, dMat2, method="ls")
fit2.ls <- contrasts.fit(fit2.ls, cMat2)

# Moderation of standard errors using empirical Bayes for first model fit
eb.ls <- eBayes(fit.ls, proportion=0.01)
eb.ls

# Moderation of standard errors using empirical Bayes for second model fit
eb2.ls <- eBayes(fit2.ls, proportion=0.01)
```

**Save the model**

With the following you can write the complete linear model analysis results to a file in a tabular format
```{r saveLmFit,eval=TRUE}
write.fit(eb.ls, file="./text/anovaTable.txt")
```

You can retrieve the top differentially expressed genes for a specific contrast in the contrast matrix by name
```{r retrieveTopDEGenes1,comment=NA}
tG <- topTable(eb.ls, coef="CellType", number=10, resort="logFC",genelist=rownames(eb.ls))
tG
tG <- tG[order(tG$ID),]
#tG <- tG[order(rownames(tG)),]

```

You can retrieve the top differentially expressed genes for a specific contrast in the contrast matrix by index
```{r retrieveTopDEGenes2,comment=NA}
tG2 <- topTable(eb2.ls, coef=1, number=10, resort="logFC",genelist=rownames(eb2.ls))
tG2
tG2 <- tG2[order(tG2$ID),]
#tG2 <- tG2[order(rownames(tG2)),]
```

The same genes were identified
```{r checkIfSameGenesWereIdentified,comment=NA}
all(tG == tG2)

# We can also return ALL the features investigated by using Inf
tG2 <- topTable(eb2.ls, coef=1, number=Inf, resort="logFC",genelist=rownames(eb2.ls))

# As you can see we need to add gene annotation!
str(tG2)
```

Retrieve and add the annotation
-------------------------

Retrieve and add gene SYMBOLS, ENTREZID, and GENENAME

Using the "old" hard-way method based on mget():
```{r mGetAnnotation}
# Get ENTREZID
egid <- mget(tG2$ID, hgu95aENTREZID, ifnotfound=NA)

# Get Gene Symbol
sym <- mget(tG2$ID, hgu95aSYMBOL, ifnotfound=NA)

# Get Gene Name
nm <- mget(tG2$ID, hgu95aGENENAME, ifnotfound=NA)
```

Add the annnotation to differential gene expression results
```{r addAnnotationToDGEResults,comment=NA}
tG2ann <- cbind(EGID=unlist(egid), SYMBOL=unlist(sym), GeneName=unlist(nm),
           tG2, stringsAsFactors=FALSE)
```

Let's check the results now

Print and see what the what the data frame looks like after adding the annotation.
```{r checkObjectsAfterAddingAnnotation,eval=FALSE,comment=NA}
tG2ann

str(tG2ann)
```


### Retrieve and add gene SYMBOLS, ENTREZID, and GENENAME

Using the "new" and simple method based on select():

```{r retrieveAnnotation,comment=NA}
### Alternative ###
### ann <- select(x=hgu95a.db, keys=rownames(tG2),cols=c("ENTREZID", "SYMBOL", "GENENAME"))

ann <- select(x=hgu95a.db, keys=tG2$ID,columns=c("ENTREZID", "SYMBOL", "GENENAME"))

# Check the dimensions of the annotation the results data.frames
dim(ann)
dim(tG2)
```

Add the annnotation to differential gene expression results

```{r addAnnotationToDGE,comment=NA}
# tG2$ID = rownames(tG2)
tG2ann2 <- merge(x=ann, y=tG2, by.x="PROBEID", by.y="ID", all.x=TRUE, all.y=FALSE)

### Let's check the results now
str(tG2ann2)
```

We can save the results as usual for later use
```{r saveResultsForLaterUse, eval = TRUE,comment=NA}
save(tG2, fit2.ls, eb2.ls, dMat2, cMat2, file="./objs/linearModel.rda")
```


We can also filter the genes and create a report page
```{r reorderGenes,comment=NA}
#tmp <- tG2[ tG2$B > 10 , ]
tmp <- tG2ann2[ tG2ann2$adj.P.Val < 1e-3 , ]


### Reorder the columns to have ENTREZID as the first column: hyperlinks to the NCBI ENTREZ GENE database will be generated
tmp <- tmp[ , c(2, 1, 3:ncol(tmp)) ]
```

Write the results to an html file
```{r htmlPage,eval = TRUE,comment=NA}

htmlpage(as.data.frame(tmp[ , 1]),
			     filename="./html/result.html",
			      othernames=as.data.frame(tmp[ , 2:ncol(tmp) ]),
			       table.head=colnames(tmp), digits=4)
```


A few useful plots
--------------------


We can make a **volcano plot**

```{r volcanoPlot,fig.height=6,fig.width=6,comment=NA}
volcanoplot(eb2.ls)
```

We can Retrieve the original gene expression for the interesting genes

All the expression values
```{r filterExpressionValues,comment=NA}
mat <- exprs(dat.rma)
dim(mat)

# Filter to the interesting ones (differentially expressed genes)
mat <- mat[rownames(mat)%in%tmp$PROBEID,]
dim(mat)
```

Generate a **heatmap()**
```{r heatMap,fig.height=6,fig.width=6,comment=NA}
heatmap(mat, scale='none', na.rm=TRUE, margins=c(15,7),
	     distfun=function(x) {dist(x, method='euclidian')},
	     hclustfun=function(x) {hclust(x, method='average')},
	     ColSideColors=as.character(2+as.numeric(pheno$CellType)))


# Add symbol as rownames using mget()
colnames(mat) <- pheno$CellLine
mySym <- mget(rownames(mat), hgu95aSYMBOL, ifnotfound=NA)
rownames(mat) <- unlist(mySym)


# Generated the annotated heatmap with heatmap()
heatmap(mat, scale='row', na.rm=TRUE, margins=c(15,7),
	     distfun=function(x) {dist(x, method='euclidian')},
	     hclustfun=function(x) {hclust(x, method='average')},
	     ColSideColors=as.character(2+as.numeric(pheno$CellType)))
```


### Session information

This is always important for debugging and posting on forums if required. 

```{r sessionInfo,comment=NA}
sessionInfo()
```
