__Author: Nitesh Turaga, May 6th 2014__

Goal: from Affymetrix raw data stored in CEL files to differential gene expression
========================================================

### Some Basics

Getting  the working directory, in my case

```{r getwd,comment=NA}
getwd()
```

Setting the working directory, in my case:
```{r setwd,eval=TRUE,comment=NA}
setwd("~/Documents/GeneExpressionDataAnalysis/affyData/")
````

Check the date

```{r checkDate,comment=NA}
date()
```

Clear the workspace, note the two "embedded" functions
```{r clearWorkSpace,comment=NA}
rm(list=ls())
```

Bioconductor
--------------
We are going to use the following packages from bioconductor:

1. affy 
2. limma 
3. hgu95a.db
4. annotate

### Load the libraries needed for the analysis,

```{r loadLibraries,comment=NA}
require(affy)
require(limma)
require(hgu95a.db)
require(annotate)
```

 Load  data previously stored
```{r loadData,comment=NA}
# Gene expression data
load("./objs/affyData.rda")

# Load  linear model results
load("./objs/linearModel.rda")
```


Create Functional Gene Set
---------------------------

Lets create the Functional Gene Set (FGS) list. We will get the data from the annotation
medata package **hgu95a.db** . Use the package name as a function for listing the annotation content

```{r FGS,comment=NA}
hgu95a()
```

Each environment in a metadata package maps identifiers to annotation information.
The annotation information can be gene/identifiers centered, or annotation centered.

Example: __"ID -> SYMBOL"__ is gene centered,
while __"PATHID -> genes"__ is annotation centered.
We need the second type of mapping: one FGS to all the genes of the FGS


This extracts mapping between kegg and affy identifiers for this platform
```{r extractMapping,comment=NA}
kegg <- as.list(hgu95aPATH2PROBE)
```

This for GENE ONTOLOGY
```{r geneOntology,comment=NA}
go <- as.list(hgu95aGO2ALLPROBES)
```

The code above created a list where each element has name
(the kegg identifiers) and contains a vector of identifiers, see below,
```{r explainList,comment=NA}
class(kegg)
length(kegg)
str(head(kegg))
```


Wilcoxon Rank-Sum Test
--------------------------

To run a Wilcoxon rank-sum test we can use the **geneSetTest()** function in limma
we need to identify the genes in the FGS, and give also a ranking statistics.
To this end we will use the t-statistics obtained using the topTable() function

The **geneSetTest()** tests whether genes in the set are more differentially expressed than genes not in the set. 

By contrast, a self-contained gene set test such as **roast()** tests whether genes in the test set are differentially expressed, in an absolute sense, without regard to any other genes on the array.



The test on the fisrt FGS: default parameters
```{r wilcoxonRankSumTest,comment=NA}
geneSetTest(tG2$ID%in%kegg[[1]],tG2$t)
```

The test on the fisrt FGS: shift toward down-reguated genes
```{r FGSdownRegulated,comment=NA}
geneSetTest(tG2$ID%in%kegg[[1]],tG2$t,alternative="down")
```

The test on the fisrt FGS: shift toward up-reguated genes
```{r FGSupregulated,comment=NA}
geneSetTest(tG2$ID%in%kegg[[1]],tG2$t,alternative="up")
```


### Visual representation

```{r barCodePlot,comment=NA}
barcodeplot(tG2$t,tG2$ID%in%kegg[[1]])
```

The test on all KEGG  FGS (after removing NA)
```{r testAllFGS,comment=NA}
kegg2 <- kegg[!is.na(kegg)]
gse <- vector()
for (i in 1:length(kegg2)){
  gse[i] <- geneSetTest(tG2$ID%in%kegg2[[i]],tG2$t)
	names(gse)[i] <- names(kegg2)[i]
}

### This is an alternative
tmp <- sapply(kegg2, function(gs, stat, ids) {
	geneSetTest(ids%in%gs,stat)
}, stat=tG2$t, ids=tG2$ID)
```

### See the results
```{r checkResults,comment=NA}
str(gse)
head(gse)
```

### The test on all KEGG  FGS (we can also test for NA inside the loop)
```{r removeNAInsideLoop,comment=NA}
gse <- vector()
for (i in 1:length(kegg)){
  #check if all values are NOT true
	if (all(!tG2$ID%in%kegg[[i]])) {
		gse[i] <- NA
	}else{
		gse[i] <- geneSetTest(tG2$ID%in%kegg[[i]],tG2$t)
	}
	names(gse)[i] <- names(kegg)[i]
}
```


Check how the gse vector looks, the length and the structure
```{r checkLength,comment=NA}
length(gse)
str(gse)
```

This is the way we can **visualize** this: **the first gene set**
```{r firstGeneSet,comment=NA}
i=10
barcodeplot(tG2$t,tG2$ID%in%kegg[[i]])
```

This is the way we can visualize this: **the smallest pvalue**
```{r visualizeSmallestPValue,comment=NA}
i=which.min(gse)
barcodeplot(tG2$t,tG2$ID%in%kegg[[i]])
```

This is the way we can visualize this: **the largest pvalue**
```{r visualizeLargestPValue,comment=NA}
i=which.max(gse)
barcodeplot(tG2$t,tG2$ID%in%kegg[[i]])
```


We can also use the geneSetTest() function differently,
you can learn about this function by typing
```{r eval=FALSE,comment=NA}
?geneSetTest

# Or simply print the arguments of the function
args(geneSetTest)
```


**We can compare our FGS to randomly generated lists of genes**
**of the same size of our FGS of interest and compute the p-value**
**as the proportion of times our FGS returns a smaller p-value**
**compared to the randomly generated lists.**
```{r rankTests,comment=NA}
geneSetTest(tG2$ID%in%kegg[[1]],tG2$t,alternative="up",type="t",
	    ranks.only=FALSE,nsim=1000)

geneSetTest(tG2$ID%in%kegg[[94]],tG2$t,alternative="down",type="t",
	    ranks.only=FALSE,nsim=1000)
```



**limma** enables you to run also self-contained tests as opposed to competitive tests.
To do this you can use the **roast()** function


Extract the expression values
```{r comment=NA}
mat <- exprs(dat.rma)
```

Select gene expression for the genes in the FGS
```{r comment=NA}
sel <- rownames(mat)%in%kegg[[1]]
```


For this analysis the design matrix must already account for the contrasts
```{r comment=NA}
newDmat <- data.frame(
		      Intercept=1,
		      CellType=dMat2[,1]+dMat2[,2] + (-1* (dMat2[,3]+dMat2[,4]))
		      )
```

Run the test on one FGS: the contrast is the column index from the design matrix
```{r comment=NA}
roast( y=mat,index=sel, design=newDmat, contrast=2)
```

The test on the first five FGS for KEGG
```{r comment=NA}
gse2 <- list()
for (i in 1:5){
	sel <- rownames(mat)%in%kegg[[i]]
  gse2[[i]] <- roast(mat,index=sel,design=newDmat, contrast=2)
	names(gse2)[i] <- names(kegg)[i]
}
```

The test on the first five FGS for GO
```{r comment=NA}
gse.go <- list()
for (i in 1:5){
	sel <- rownames(mat)%in%go[[i]]
  gse.go[[i]] <- roast(mat,index=sel, design=newDmat, contrast=2)
	names(gse.go)[i] <- names(go)[i]
}
```

### Now how we do find out which patwhays are enriched?

Load the KEGG library
```{r comment=NA}
library(KEGG.db)
library(GO.db)
```

### If you do not have it install it from bioconductor
```{r comment=NA}
installedPckgs <- installed.packages()[,"Package"]
# Here we define the list of desired libraries
pckgListBIOC <- c("KEGG.db")
# We source the biocLite.R script from the Bioconductor website
source("http://bioconductor.org/biocLite.R")
# Load the packages, or install them from Bioconductor if needed
for (pckg in pckgListBIOC) {
    if (! pckg %in% installedPckgs) biocLite(pckg)
    require(pckg, character.only=TRUE)
}
require("KEGG.db")
```

Check the content of the KEGG metadata package
```{r comment=NA}
KEGG()
GO()
```

It appears that KEGGPATHID2NAME contains what we want,
so you can use mget() to obtain the patways names
```{r comment=NA}
pathNames <- mget(names(gse), KEGGPATHID2NAME, ifnotfound=NA)
goNames <- mget(names(gse.go),GOTERM,ifnotfound=NA)
goNames <- lapply(goNames,Term)
```

Print pathNames and goNames
```{r,eval=FALSE,comment=NA}
print(pathNames)
print(goNames)
```

Retrieve the smallest p-values
```{r smallestPval,comment=NA}
pathNames[names(pathNames)==names(gse[which.min(gse)])]
```

If they are in the same order
```{r,comment=NA}
pathNames[which.min(gse)]
```

Retrieve the smallest p-values
```{r,comment=NA}
pathNames2 <- pathNames[length(pathNames):1]
```

Always correct also if not in the same order
```{r comment=NA}
pathNames2[names(pathNames2)==names(gse[which.min(gse)])]
```

Wrong if they are in the same order
```{r comment=NA}
pathNames2[which.min(gse)]
```

Retrieve fgs with p-value less than 0.001
```{r comment=NA}
pathNames[gse<1e-5]
```

### Common genes
```{r comment=NA}
index <- which(gse<1e-2)
myFGS <- kegg[index]
sum(myFGS[[1]]%in%myFGS[[2]])
```

### Cross tabulation
```{r comment=NA}
all <- unique(unlist(myFGS))
table(fgs1=all%in%myFGS[[1]],fgs2=all%in%myFGS[[2]])
```

### Venn diagram
```{r comment=NA}
vennDiagram(data.frame(fgs1=all%in%myFGS[[1]],
		       fgs2=all%in%myFGS[[2]],fgs3=all%in%myFGS[[3]]))
```

### Multiple testing correction
```{r multipleTestingCorrection,comment=NA}
library(multtest)

# Correct p-values
adjp <- mt.rawp2adjp(gse)
adjp <- adjp$adjp[order(adjp$index),]

# Bonferroni only
adjp <- mt.rawp2adjp(gse,proc="Bonferroni")
adjp <- adjp$adjp[order(adjp$index),]

# Benjamini-Hochberg only
adjp <- mt.rawp2adjp(gse,proc="BH")
adjp <- adjp$adjp[order(adjp$index),]
```

Make an histogram of the p-values
```{r comment=NA}
hist(adjp[,1], nclass=20)
hist(adjp[,2],nclass=20,add=TRUE,col="blue")
```


### Session information
```{r sessionInformation,comment=NA}
sessionInfo()
```
