__Author: Nitesh Turaga, May 6th 2014__


Goal: Gene Set Analysis
========================================================

### Some Basics

Getting  the working directory, in my case

```{r getwd}
getwd()
```

Setting the working directory, in my case:
```{r setwd,eval=TRUE}
setwd("~/Documents/GeneExpressionDataAnalysis/GeneSetAnalysis/")
````

Check the date

```{r checkDate}
date()
```

Clear the workspace, note the two "embedded" functions
```{r clearWorkSpace}
rm(list=ls())
```

Bioconductor
-------------

The following R packages were used to perform our analyses 

1. **BiocGenerics**: this library implements several generic classes and methods to work with {\Bioc} packages and can be obtained from Bioconductor.

2. **Biobase**: this library implements classes and methods to work with genomic data and can be obtained from {\Bioc};

3. **limma**: this library provide advanced methods to perform gene expression analysis from raw data to gene set analysis and can be obtained from Bioconductor.

4. **RTopper**: this library implements methods to perform integrated gene set analysis across datasets and platforms and can be obtained from Bioconductor.

5. **multtest**: this library implements methods to perform multiple test correction and can be obtained from Bioconductor.

6. **org.Hs.eg.db**: this library contains human genes annotation and can be obtained from Bioconductor.

7. **KEGG.db**: this library contains KEGG pathway annotation and can be obtained from Bioconductor.

8. **reactome.db**: this library contains Reactome pathway annotation and can be obtained from Bioconductor.

9. **GO.db**: this library contains Gene Ontoogy annotation and can be obtained from Bioconductor.

10. **AnnotationDBI** and **annotate**: these libraries contain classes and methods to access and manipulated annotation packages and can be obtained from Bioconductor.



Source the biocLite.R script from Bioconductor

```{r installBiocPackages}
source("http://bioconductor.org/biocLite.R")
###Get the list of available packages
installedPckgs <- installed.packages()[,"Package"]
###Define the list of desired libraries
pckgListBIOC <- c("BiocGenerics", "Biobase", "limma", "RTopper", 
  	  "org.Hs.eg.db", "AnnotationDbi", "annotate", 
		  "multtest", "KEGG.db", "GO.db")
###Load the packages, install them from Bioconductor if needed
for (pckg in pckgListBIOC) {
	if (! pckg %in% installedPckgs) {
		biocLite(pckg, suppressUpdates=TRUE, ask=FALSE)
		}
	require(pckg, character.only=TRUE)
}
```

### code chunk number 3: Load Data and packages required
```{r}
require(RTopper)
data(sepScores)
ls()
class(sepScores)
names(sepScores)
sapply(sepScores, class)
sapply(sepScores,dim)
```

### code chunk number 4: metaData
```{r load annotation database}
require(org.Hs.eg.db)
org.Hs.eg()
```


### code chunk number 5: list FGS (Functional Gene Set)
```{r}
kegg <- as.list(org.Hs.egPATH2EG)
length(kegg)
str(kegg[1:5])
names(kegg)[1:5]
go <- as.list(org.Hs.egGO2ALLEGS)
length(go)
str(go[1:5])
names(go)[1:5]
```

### code chunk number 6: convertIDs
```{r convertIDs}
numberOfFGSkegg <- 200
kegg <- lapply(kegg[sample(1:length(kegg),numberOfFGSkegg)],
               function(x) unique(unlist(mget(x,org.Hs.egSYMBOL))))
str(kegg[1:5])
### Process GO: keep only Biological Process
length(go)
go <- go[ Ontology(names(go)) == "BP" ]
length(go)
numberOfFGSgo <- 200
go <- lapply(go[sample(1:length(go),numberOfFGSgo)],
             function(x) unique(unlist(mget(x,org.Hs.egSYMBOL))))
str(go[1:5])
```


### code chunk number 7: annotateFGS
```{r annotateFGS}
require(KEGG.db)
KEGG()
names(kegg) <- paste(names(kegg),unlist(mget(names(kegg),KEGGPATHID2NAME)),sep=".")
head(names(kegg), n=10)
```

### code chunk number 8: listFGS
```{r ListFGS}
require(GO.db)
GO()
names(go) <- paste(names(go),Term(names(go)),sep=".")
head(names(go), n=10)
```

### code chunk number 9: listFGS
```{r}
fgsList <- list(go=go,kegg=kegg)
str(fgsList$go[1:5])
str(fgsList$kegg[1:5])
```

### code chunk number 10: runGSEbatchArgs
```{r}
args(runBatchGSE)
```


### code chunk number 11: runBatchGSE.separate
```{r}
gseABS.sep <- runBatchGSE(dataList=sepScores, fgsList=fgsList)
gseABS.sep <- runBatchGSE(dataList=sepScores, fgsList=fgsList,
				 absolute=TRUE, type="f", alternative="mixed")
```

### code chunk number 12: runBatchGSE.separate2
```{r}
gseUP.sep <- runBatchGSE(dataList=sepScores, fgsList=fgsList,
				 absolute=FALSE, type="t", alternative="up")
gseDW.sep <- runBatchGSE(dataList=sepScores, fgsList=fgsList,
				 absolute=FALSE, type="t", alternative="down")
gseBOTH.sep <- runBatchGSE(dataList=sepScores, fgsList=fgsList,
				 absolute=FALSE, type="t", alternative="either")
```

### code chunk number 13: runBatchGSE.int3
```{r}
gseABSsim.sep <- runBatchGSE(dataList=sepScores, fgsList=fgsList,
				    absolute=TRUE, type="f", alternative="mixed",
				    ranks.only=FALSE, nsim=1000)
gseUPsim.sep <- runBatchGSE(dataList=sepScores, fgsList=fgsList,
				    absolute=FALSE, type="t", alternative="up",
				    ranks.only=FALSE, nsim=1000)
```

### code chunk number 14: runBatchGSE.format1
```{r}
str(gseUP.sep[1:5])
head(gseABSsim.sep$dat.affy$go)
head(gseABSsim.sep$dat.affy$kegg)
```

### code chunk number 15: runBatchGSE.altFunc
```{r}
require(limma)
gseUP.sep.2 <- runBatchGSE(dataList=sepScores, fgsList=fgsList,
				 absolute=FALSE, gseFunc=wilcoxGST, alternative="up")
```

### code chunk number 16: runBatchGSE.format2
```{r}
str(gseUP.sep.2)
all(gseUP.sep.2$go==gseUP.sep$go)
```

### code chunk number 17: runBatchGSE.altFunc2
```{r}
gseFunc <- function (selected, statistics, threshold) {
	diffExpGenes <- statistics > threshold
	tab <- table(diffExpGenes, selected)
	pVal <- fisher.test(tab)[["p.value"]]
	}
gseUP.sep.3 <- runBatchGSE(dataList=sepScores, fgsList=fgsList,
				 absolute=FALSE, gseFunc=gseFunc, threshold=7.5)
```

### code chunk number 18: runBatchGSE.format3
```{r}
str(gseUP.sep.3)
head(data.frame(fisher=gseUP.sep.3$dat.affy$kegg,
                wilcoxon=gseUP.sep$dat.affy$kegg))
```

### code chunk number 19: adjustP
```{r}
gseABS.sep.BH <- adjustPvalGSE(gseABS.sep)
gseABS.sep.holm <- adjustPvalGSE(gseABS.sep, proc = "Holm")
```

### code chunk number 20: adjusted.format
```{r}
names(gseABS.sep.BH)
names(gseABS.sep.holm)
str(gseABS.sep.BH$dat.affy)
str(gseABS.sep.holm$dat.affy)
head(gseABS.sep.BH$dat.affy$go, n=10)
head(gseABS.sep.holm$dat.affy$go, n=10)
```

### code chunk number 21: figure004
```{r}
hist(gseABS.sep.BH$dat.affy$go[ , "rawp"], nclass=10, col="orange",
     main="P-values")
```

### code chunk number 22: figure005
```{r}
hist(gseABS.sep.BH$dat.affy$go[ , "BH"], nclass=10, col="orange",
     main="Q-values")
```


### code chunk number 23: readAndProcessData
```{r}
autismGenes <- read.table("myData/101symbols_v02.txt", sep="\t", header=TRUE,
                          colClasses="character")
str(autismGenes)
```

### code chunk number 24: newFunctionForLists
```{r}
formatGeneList <- function (geneList, allGenes) {
  stats <- 1* allGenes %in% geneList
  names(stats) <- allGenes
  out <- list(GeneList = stats)
	}
```

### code chunk number 25: formatTheGeneLsit
```{r}
allGenes <- unique(unlist(as.list(org.Hs.egSYMBOL)))
str(allGenes)
geneMembership <- formatGeneList(autismGenes$SYMBOL, allGenes)
str(geneMembership)
table(geneMembership$GeneList)
```

### code chunk number 26: runBatchGSE.altFunc3
```{r}
gseMyList <- runBatchGSE(dataList=geneMembership, fgsList=fgsList,
				 absolute=FALSE, gseFunc=gseFunc, threshold=0.5)
```

### code chunk number 27: showTheResults
```{r}
str(gseMyList)
str(gseMyList$GeneList$kegg)
summary(gseMyList$GeneList$kegg)
table(is.na(gseMyList$GeneList$kegg))
table(gseMyList$GeneList$kegg < 0.01)
table(gseMyList$GeneList$go < 0.01)
```

### code chunk number 28: adjustP.2
```{r}
gseMyList.BH <- adjustPvalGSE(gseMyList)
gseMyList.BH$GeneList$go[ gseMyList.BH$GeneList$go[, "BH"] < 0.01 , ]
gseMyList.BH$GeneList$kegg[ gseMyList.BH$GeneList$kegg[, "BH"] < 0.01 , ]
```

### code chunk number 29: sessioInfo
```{r}
sessionInfo()
```

